docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,230 @@
+
+Warm nodes - The concept of keeping up extraneous and unused nodes to prevent having to wait for patches or security updates, thus allowing you to have those nodes ready when more users try to use your service without the wait times.
+
+Shared Image Gallery - A repository that is an Azure Resource for managing and sharing images
+
+Snapshot - A static copy of the image at the point in time that it is taken
+
+Blessed Node - Whichever node has been the most stable and is chosen off of that metric as the one to clone the OS image of and cache it for future nodes to use
+
+## Summary
+
+A controller will be added to CAPZ which caches the Nodes’ OS image on a regular interval and updates the model to use that image for future scale outs. This feature will allow for:
+- faster horizontal scaling 
+- prevent security alerts from new nodes which immediately need a security update as those new nodes will now come with the security update already installed
+- help prevent users from needing to spin up warm nodes and overprovision
+
+## Motivation
+
+A model scenario would be an operator spinning up a CAPZ cluster and having this feature be able to be toggled on or off with an environment variable. If it was toggled on, then as the months passed and more security updates and patches needed to be applied to the operator’s node OS image, these changes would be cached on a regular interval and the operator would no longer have to wait for these changes to apply on new node creations. As a result, users will have faster horizontal scaling and require fewer warm nodes and overprovisioning to avoid this problem, especially since the new nodes will have the container images of the applications it will run pre-cached so pods will run quicker when scheduled. This feature will also help users have better security compliance as new nodes will already be compliant instead of needing to patch.
+
+#### Credits
+
+A big thank you to Jack Francis, Michael Sinz, and Amr Hanafi for their work on the Kamino project that used a similar Prototype Pattern with AKSEngine which inspired this feature request: https://github.com/jackfrancis/kamino
+
+### Goals
+
+1. A solution using the Node Prototype pattern which cache’s nodes’ OS image on a regular interval and replaces it as a new ‘golden image’.
+1. Faster horizontal scale outs
+1. Prevent security breaking issues on node bootup from security updates being required immediately
+
+We will know we’ve succeeded when we can benchmark speed increases and successful image changes
+
+### Non-Goals/Future Work
+
+1. Extend the functionality to Windows nodes
+1. Optimization for efficiency and scalability
+1. A more complicated method of selecting a good candidate node such as incorporating manual forced prototype creation, which would perform better but take research to find the optimal method. 
```

Comment:
nit: some bullet points have "." at the end and others don't, let's change to one or the other to be consistent

Reply:
Consistently no "." at the end for the Goals/Non-Goals/Future work section